pycbc/events/stat.py

```diff
@@ -1741,6 +1743,39 @@ def logsignalrate(self, stats, shift, to_shift):
         logr_s += numpy.log((self.curr_mchirp / 20.) ** (11. / 3.))
         return logr_s
 
+    def single(self, trigs):
+        """
+        Calculate the necessary single detector information
+
+        In this case the ranking rescaled (see the lognoiserate method here)
+        with the phase, end time, sigma, SNR, template_id and the
+        benchmark_logvol values added in. This also stored the current chirp
+        mass for use when computing the coinc statistic values.
+
+        Parameters
+        ----------
+        trigs: dict of numpy.ndarrays, h5py group or similar dict-like object
+            Object holding single detector trigger information.
+
+        Returns
+        -------
+        numpy.ndarray
+            The array of single detector values
+        """
+        from pycbc.conversions import mchirp_from_mass1_mass2
+        try:
+            mass1 = trigs.param['mass1']
+            mass2 = trigs.param['mass2']
+        except AttributeError:
+            mass1 = trigs['mass1']
+            mass2 = trigs['mass2']
+        self.curr_mchirp = mchirp_from_mass1_mass2(mass1, mass2)
+
+        if self.mcm is not None:
+            # Careful - input might be a str, so cast to float
+            self.curr_mchirp = min(self.curr_mchirp, float(self.mcm))
```

Comment:
This method might be time-critical. Can the cast to float be done during initialization instead?

Reply:
More generally, I would consider it a bug if `self.mcm` has a string type. A chirp mass is conceptually a positive real number, so a variable representing a chirp mass should always have a real numeric type. What is causing `self.mcm` to be a string?

Reply:
Re-ordering the methods in the class and creating `rank_stat_coinc` has caused this to appear as though i've created the `single` method whereas I have just re-ordered them, unsure how to get that to appear properly.

The `if self.mcm is not None:` was added in this PR: https://github.com/gwastro/pycbc/pull/3009

So it looks like `self.mcm` might be a str type in offline?

Reply:
I would check this with Tom/Ian. I do not understand #3009, it was not properly documented and it looks like a mistake on a cursory look.

Reply:
I will copy and paste @spxiwh's comment from my asking him:
"My hunch is that this is stored in the config file as a string for offline, and nothing outside of the stat module would, or should, know that, and so the top-level code would just look for something like an “additional kwargs” section and simply pass the entire contents of that to the stat class. The stat class would then need to parse that (which in this case would mean translating “65" -> 65). Live works quite differently, and doesn’t use config files. If it’s having to hardcode any possible statistic key-word argument on the command-line (perhaps not ideal) then it would enforce type at that stage."

I agree with you that the float type casting doesn't need to be done every single time this method is called so will move it to the initialisation.

Reply:
Ok I see. Well, if this turns out to be a problem (either performance-wise, or type incompatibility) we will find out pretty soon anyway.

Reply:
`self.mcm` will be None in pycbc live so this if statement will execute, I'm prepared to leave it be if that's okay